tests/FsCodec.SystemTextJson.Tests/UnionConverterTests.fs

```diff
@@ -0,0 +1,361 @@
+    | CaseT (Some x, y) -> sprintf """{"case":"CaseT","a":"%s","b":"%s"}""" x.Value y.Value
+    | CaseU skus -> sprintf """{"case":"CaseU","Item":[%s]}""" (skus |> Seq.map (fun s -> sprintf "\"%s\"" s.Value) |> String.concat ",")
+    | CaseV skus -> sprintf """{"case":"CaseV","skus":[%s]}""" (skus |> Seq.map (fun s -> sprintf "\"%s\"" s.Value) |> String.concat ",")
+    | CaseW (id, skus) -> sprintf """{"case":"CaseW","Item1":"%s","Item2":[%s]}""" id.Value (skus |> Seq.map (fun s -> sprintf "\"%s\"" s.Value) |> String.concat ",")
+    | CaseX (id, skus) -> sprintf """{"case":"CaseX","a":"%s","skus":[%s]}""" id.Value (skus |> Seq.map (fun s -> sprintf "\"%s\"" s.Value) |> String.concat ",")
+    | CaseY (a, b) -> sprintf """{"case":"CaseY","a":"%s","b":"%s"}""" (string a) (string b)
+    | CaseZ (a, None) when ignoreNulls -> sprintf """{"case":"CaseZ","a":"%s"}""" (string a)
+    | CaseZ (a, None) -> sprintf """{"case":"CaseZ","a":"%s","b":null}""" (string a)
+    | CaseZ (a, Some b) -> sprintf """{"case":"CaseZ","a":"%s","b":"%s"}""" (string a) (string b)
+    | CaseZ (a, Some b) -> sprintf """{"case":"CaseZ","a":"%s","b":"%s"}""" (string a) (string b)
+
+type FsCheckGenerators =
+    static member CartId = Arb.generate |> Gen.map CartId |> Arb.fromGen
+    static member SkuId = Arb.generate |> Gen.map SkuId |> Arb.fromGen
+
+type DomainPropertyAttribute() =
+    inherit FsCheck.Xunit.PropertyAttribute(QuietOnSuccess = true, Arbitrary=[| typeof<FsCheckGenerators> |])
+
+let roundtripProperty ignoreNulls (profile : JsonSerializerOptions) value =
+    let serialized = JsonSerializer.Serialize(value, profile)
+    render ignoreNulls value =! serialized
+    let deserialized = JsonSerializer.Deserialize<_>(serialized, profile)
+    deserialized =! value
+
+let includeNullsProfile = Options.Create(ignoreNulls = false)
+[<DomainPropertyAttribute(MaxTest=1000)>]
+let ``UnionConverter ignoreNulls Profile roundtrip property test`` (x: TestDU) =
+    let ignoreNulls, profile = false, includeNullsProfile
+    profile.IgnoreNullValues =! false
+    roundtripProperty ignoreNulls profile x
+
+let defaultProfile = Options.Create ()
+[<DomainPropertyAttribute(MaxTest=1000)>]
+let ``UnionConverter opinionated Profile roundtrip property test`` (x: TestDU) =
+    let ignoreNulls, profile = false, defaultProfile
+    profile.IgnoreNullValues =! false
+    roundtripProperty ignoreNulls profile x
+
+//[<Fact>]
+//let ``Implementation ensures no internal errors escape (which would render a WebApi ModelState.Invalid)`` () =
+//    let s = JsonSerializer.CreateDefault()
+//    let mutable gotError = false
+//    s.Error.Add(fun _ -> gotError <- true)
+
+//    let dJson = """{"case":"CaseD","a":"hi"}"""
+//    use dReader = new StringReader(dJson)
+//    use dJsonReader = new JsonTextReader(dReader)
+//    let d = s.Deserialize<TestDU>(dJsonReader)
+
+//    test <@ (CaseD "hi") = d @>
+//    test <@ false = gotError @>
```

Comment:
TODO: I'll need to poke around to see how they have integrated STJ with ASP.NET's modelstate to make sure I can test for this behaviour. (JsonSerializer.Error is not a thing).